hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieHadoopFsRelationFactory.scala

```diff
@@ -330,7 +338,64 @@ class HoodieMergeOnReadIncrementalHadoopFsRelationFactoryV2(override val sqlCont
                                                             isBootstrap: Boolean,
                                                             rangeType: RangeType = RangeType.OPEN_CLOSED)
   extends HoodieMergeOnReadIncrementalHadoopFsRelationFactory(sqlContext, metaClient, options, schemaSpec, isBootstrap,
-    MergeOnReadIncrementalRelationV2(sqlContext, options, metaClient, schemaSpec, None, rangeType))
+    MergeOnReadIncrementalRelationV2(sqlContext, options, metaClient, schemaSpec, None, rangeType)) {
+
+  private lazy val extendedSchemaSpec: Option[StructType] = {
+    if (!metaClient.isMetadataTable) {
+      val baseSchema = schemaSpec.getOrElse(tableStructSchema)
+      Some(extendSchemaForCompletionTime(baseSchema))
+    } else {
+      schemaSpec
+    }
+  }
+
+  private lazy val incrementalFileIndexV2 = new HoodieIncrementalFileIndex(
+    sparkSession, metaClient, extendedSchemaSpec, options, fileStatusCache, true,
+    MergeOnReadIncrementalRelationV2(sqlContext, options, metaClient, extendedSchemaSpec, None, rangeType))
+
+  override def buildFileIndex(): HoodieFileIndex = incrementalFileIndexV2
+
+  override def buildDataSchema(): StructType = {
+    val baseSchema = super.buildDataSchema()
+    extendSchemaForCompletionTime(baseSchema)
+  }
+
+  override def buildFileFormat(): FileFormat = {
+    val tableConfig = metaClient.getTableConfig
+    val (finalAvroSchema, finalStructSchema) = if (!metaClient.isMetadataTable) {
+      (extendAvroSchemaForCompletionTime(tableAvroSchema), buildDataSchema())
+    } else {
+      (tableAvroSchema, buildDataSchema())
+    }
+
+    new HoodieFileGroupReaderBasedFileFormat(basePath.toString,
+      HoodieTableSchema(finalStructSchema, finalAvroSchema.toString, internalSchemaOpt),
+      tableConfig.getTableName, queryTimestamp.get, getMandatoryFields, isMOR, isBootstrap,
+      isIncremental, validCommits, shouldUseRecordPosition, getRequiredFilters,
+      tableConfig.isMultipleBaseFileFormatsEnabled, tableConfig.getBaseFileFormat)
+  }
+
+  private def extendSchemaForCompletionTime(baseSchema: StructType): StructType = {
+    val completionTimeField = StructField(HoodieRecord.COMMIT_COMPLETION_TIME_METADATA_FIELD, StringType, nullable = true)
+    StructType(baseSchema.fields :+ completionTimeField)
+  }
+
+  private def extendAvroSchemaForCompletionTime(baseAvroSchema: Schema): Schema = {
```

Comment:
This repeats code in another section of this PR. Can we move this into a common utility?